### DISCUSSION TITLE
Conceptual question about scorer

### DISCUSSION THREAD

**Discussion by fquirin** [coqui-ai/STT#2217] [Q&A]
Hello everybody,

I have a conceptual question about the LM aka scorer in Coqui. 

A bit of background first. I've been working on open-source voice assistants for a long time now (e.g. [SEPIA](https://github.com/SEPIA-Framework)) and although the ASR systems have come a long way (since Sphinx4 ^^) it is still necessary to build small domain, custom LMs to get acceptable recognition quality.
Pre-trained models usually come with larger LMs and since it's often not practicable or possible to reuse the original training data I have to replace the original LM completely instead of simply "augmenting" it with my own data.

So here is my question. When I noticed that the Coqui models were performing quite OK even without a scorer I was wondering if it's possible to use a custom scorer, trained on a few dozens of own sentences and apply it in a way so that Coqui is preferring those sentences without loosing the original large vocabulary completely? Basically just shifting weights a bit?
I have a feeling that the alpha and beta arguments for the scorer might be able to control this but I can't find any explanation of what they actually do.

[EDIT] The 'hot words' feature seems to be very similar, but boosts only single words not sentences

Thanks in advance for any info or help :slightly_smiling_face: 
Florian

**Comment by wasertech** [comment]
You are correct. Building your own little scorer for your task with a pre-trained model is the fastest way the improve your results. 
There is practically no reason to fine-tune the acoustic models (they are quite on spot).
Try to aim for a scorer under 800 Mb.
To create your source_lm.txt file, try to use around 40 Go of normalized text and it should be good to optimize.
Use the lm_optimize python script to find the best values for alpha and beta so you can export with them, this will make `$LM_N_TRIALS` test to hyper-optimize the value of alpha and beta within the specified max range `$LM_ALPHA_MAX` and `$LM_BETA_MAX`.
As for word boost it can help here and there if you say a word a lot and it doesn't pick it just right but the word must be present in the language model to begin with.

```bash
# Make a dummy scorer to find apha and beta
generate_lm.py \
		--input_txt /mnt/extracted/sources_lm.txt \
		--output_dir /mnt/lm/ \
		--top_k ${LM_TOP_K} \
		--kenlm_bins ${HOMEDIR}/kenlm/build/bin/ \
		--arpa_order 4 \
		--max_arpa_memory "85%" \
		--arpa_prune "0|0|1" \
		--binary_a_bits 255 \
		--binary_q_bits 8 \
		--binary_type trie
generate_scorer_package \
		--checkpoint /mnt/models/ \
		--lm /mnt/lm/lm.binary \
		--vocab /mnt/lm/vocab-${LM_TOP_K}.txt \
		--package /mnt/lm/kenlm.scorer \
		--default_alpha 0.0 \
		--default_beta 0.0
# Find best values
lm_optimizer.py \
		--show_progressbar true \
		--train_cudnn true \
		--alphabet_config_path /mnt/models/alphabet.txt \
		--scorer_path /mnt/lm/kenlm.scorer \
		--feature_cache /mnt/sources/feature_cache \
		--test_files ${all_test_csv} \
		--test_batch_size ${TEST_BATCH_SIZE} \
		--n_hidden ${N_HIDDEN} \
		--lm_alpha_max ${LM_ALPHA_MAX} \
		--lm_beta_max ${LM_BETA_MAX} \
		--n_trials ${LM_N_TRIALS} \
		--checkpoint_dir /mnt/checkpoints/
# Repackage the scorer with the correct values
rm  /mnt/lm/kenlm.scorer
generate_scorer_package \
		--checkpoint /mnt/models/ \
		--lm /mnt/lm/lm.binary \
		--vocab /mnt/lm/vocab-${LM_TOP_K}.txt \
		--package /mnt/lm/kenlm.scorer \
		--default_alpha ${LM_ALPHA} \
		--default_beta ${LM_BETA}
# Testing with the best values
python -m coqui_stt_training.evaluate \
		--show_progressbar true \
		--train_cudnn true \
		${AMP_FLAG} \
		--alphabet_config_path /mnt/models/alphabet.txt \
		--scorer_path /mnt/lm/kenlm.scorer \
		--test_files ${all_test_csv} \
		--test_batch_size ${TEST_BATCH_SIZE} \
		--n_hidden ${N_HIDDEN} \
		--lm_alpha ${LM_ALPHA} \
		--lm_beta ${LM_BETA} \
		--checkpoint_dir /mnt/checkpoints/ \
		--test_output_file /mnt/models/test_output.json
# Exporting the best models
/tflite-venv/bin/python -m coqui_stt_training.export \
		--alphabet_config_path /mnt/models/alphabet.txt \
		--scorer_path /mnt/lm/kenlm.scorer \
		--feature_cache /mnt/sources/feature_cache \
		--n_hidden ${N_HIDDEN} \
		--beam_width ${BEAM_WIDTH} \
		--lm_alpha ${LM_ALPHA} \
		--lm_beta ${LM_BETA} \
		--load_evaluate "best" \
		${LOAD_CHECKPOINT_FROM} \
		--export_dir /mnt/models/ \
		--export_tflite true \
		${ALL_METADATA_FLAGS} \
		${METADATA_MODEL_NAME_FLAG}
```

**Reply by fquirin** [reply]
Thanks for the answer and the extensive train script :+1: ,
but I'm not sure if this answers my question, especially this part: `... so that Coqui is preferring those sentences without loosing the original large vocabulary completely?`.

> ... but the word must be present in the language model to begin with

If I understand your last sentence correct it is probably not possible?
An example:
If I train a scorer with one sentence and 10 words, will the model only be able to recognize those 10 words, no matter what alpha and beta values are used?

**Reply by wasertech** [reply]
> If I train a scorer with one sentence and 10 words, will the model only be able to recognize those 10 words, no matter what alpha and beta values are used?

Yes but without optimization you might get mis-transcriptions where some words are not picked up or swapped (it's more an issues with a real - useful - scorer) (you'll probably also could transcribe only this exact sentence).

> Pre-trained models usually come with larger LMs and since it's often not practicable or possible to reuse the original training data I have to replace the original LM completely instead of simply "augmenting" it with my own data.

You need a lot of sentences to make a scorer useful (but compared to the acoustic model it's peanuts in size).
Still 12 or so sentences won't be useful in the real word. Try with a few million sentences? You want to be able speak lots of words in different context after all na?

> it's often not practicable or possible to reuse the original training data

Everyone should be able to reproduce a pre-trained model. If you are having trouble locating some piece of data try to reach the maintainer of the pre-trained model for sources.

Use the data from the original scorer so you can augment it with your own (try to add more than 12 sentences you can do better...). 
- They have to be close to the targeted context. 
- Try to use multiple examples to say the same thing but with different words.
- Try to use the same words in contextually different sentences.

To optimize you have to download and prepare one (or more) audio test set(s). 

Your options are:
- You can use test sets from the acoustic model to easily compare your scorer to the pre-trained one. (This will take a lot of time and space on your drives to store all the data for the acoustic model - can't only download test sets - )
- Or you can just create a small labelled audio test set for the optimisation (not containing the same sentences as the LM). (To compare both scorers you'll have to optimize them both on your test set and see if your sentences improve the results on your test)

Both case, this process is long and requires lot of computing power but the ratio `cost / reward` makes optimization the "best bang for your buck".

If after this step your model is having trouble with words you say a lot, then try to boost those "hot words".

**Reply by fquirin** [reply]
I appreciate your answer but I should probably clarify some things.

> You need a lot of sentences to make a scorer useful (but compared to the acoustic model it's peanuts in size).
Still 12 or so sentences won't be useful in the real word. Try with a few million sentences? You want to be able speak lots of words in different context after all na?

No, I don't want to do that because the real-life WER of the large LMs is still too high and in addition they often require too much RAM to work on the platforms I'm using (e.g. Raspberry Pi) ;-). That is why I drastically reduce the vocabulary and train multiple use-case specific LMs with very low WER. Up until now I've [done this](https://github.com/fquirin/kaldi-adapt-lm) for Vosk and Zamia (Kaldi) models with solid results.
When I started to experiment with the "raw" Coqui models (without scorer) I asked myself if the scorer might work in a different way so that it could be used to boost the small vocabulary without completely loosing the larger one, but this doesn't seem to be possible from what I've seen now :-/

**Reply by wasertech** [reply]
> I don't want to do that because the real-life WER of the large LMs is still too high and in addition they often require too much RAM to work on the platforms I'm using (e.g. Raspberry Pi)

My latest scorer has a billion sentences; using a top K of 500 000. Yet [it only weights 689MB](https://coqui.ai/french/commonvoice-fr/v0.9). Plenty to fit in a pi's ram.

> `Best params: lm_alpha=0.5853937631308501 and lm_beta=1.8116059936952276 with WER=0.1921812008479461`

[Average WER on my best acoustic model is under 20%](https://github.com/wasertech/commonvoice-fr/releases/tag/v0.9.0-fr-0.1).

So the hardware can do it, why can't you?

>  That is why I drastically reduce the vocabulary and train multiple use-case specific LMs with very low WER.

This technique can be useful (wakeup scorers i.e.) but has severe limitations as you can't predict the input ahead of inference.

> When I started to experiment with the "raw" Coqui models (without scorer) I asked myself if the scorer might work in a different way so that it could be used to boost the small vocabulary without completely loosing the larger one, but this doesn't seem to be possible from what I've seen now :-/

You can't "fine-tune" the scorer but if you want to try, we welcome PRs.

**Reply by fquirin** [reply]
> So the hardware can do it, why can't you?

The hardware can do it IF you run a single ASR model with a single user and have basically no other application on your Pi. This can be true, but usually I'd expect my users to have a Pi with 2-4 GB RAM, running for example a SEPIA server or client, maybe a smart-home HUB and the STT server. In the worst case you'd even get 2-3 simultaneous requests with different ASR models. So yes, it might work, but will it be safe and stable? ... depends.

> Average WER on my best acoustic model is under 20%

That is a respectable number, but it is not enough for a good user experience, sorry. Especially when you assume that the WER will be even worse under real-life conditions with noisy background or far-field speech.

> This technique can be useful (wakeup scorers i.e.) but has severe limitations as you can't predict the input ahead of inference.

I can predict the input in many cases. When a user is asking for timers, reminders, weather, smart-home controls, etc. the vocabulary isn't very big. Only in the case of translations, unspecific web searches, knowledge requests, navigation or large music libraries and obviously small-talk it gets tricky. My system can theoretically assign different (small) language models to each of these services, but it would be great to use a larger, pre-trained model where you can tweak the weights somehow for your domain without loosing all the OOV words.

> You can't "fine-tune" the scorer but if you want to try, we welcome PRs.

It would certainly be an interesting task, but I'm an application developer not an expert for basic research in speech recognition systems and my hardware capabilities are pretty limited as well.

Still I wonder, have you seen any concepts yet for such a system where a smaller LM is used on top of a good acoustic model (trained on subword tokens?) just to increase recognition in a specific domain but without restricting the vocabulary completely?

**Reply by wasertech** [reply]
> I'd expect my users to have a Pi with 2-4 GB RAM, running for example a SEPIA server or client, maybe a smart-home HUB and the STT server.

Maybe the raspberry pi isn't the best platform for your needs. 2-4 Gb of RAM is plenty but if you need more space, you can always create a swap space (it's slower than ram but it's a good trade-off if you need more memory). The CPU is the bottleneck here. The thread count to be precise (up to 8 simultaneous tasks on the beefier models). A pi is a small board computer to test and learn about computers . It's not designed to be your all in one server.

> That is a respectable number, but it is not enough for a good user experience, sorry. Especially when you assume that the WER will be even worse under real-life conditions with noisy background or far-field speech.

It's not "respectable". It's state-of-the-art for French. You need to understand that 19% average WER for an **open-domain scorer** is really hard to reach even English which has more data available doesn't come close to it.

Keep in mind that you are not doing an open domain scorer.

> I can predict the input in many cases. When a user is asking for timers, reminders, weather, smart-home controls, etc. the vocabulary isn't very big.

Since you operate within specific domains, having one specialized scorer for those could lower your average WER to less than 10% given your scorer is on point (meaning it contains all the sentences related to the domains). So I see no need to load and unload small useless scorers every time the domain changes. One scorer with all your domains included is always going to perform better just because you can infer the unexpected.

> Still I wonder, have you seen any concepts yet for such a system where a smaller LM is used on top of a good acoustic model (trained on subword tokens?) just to increase recognition in a specific domain but without restricting the vocabulary completely?

No. Not with KenLM at least. STT supports custom language models so technically nothing is stopping you to implement another LM with those desired traits.

**Reply by HarikalarKutusu** [reply]
> STT supports custom language models ...

@wasertech, can you give some pointers for this?

**Reply by wasertech** [reply]
> can you give some pointers for this?

Not really as I've never done it. 😅

I would imagine it would go something like this:

1. Find a nice paper describing a language modeling technique you like.
2. Implement it with C++ preferably and make a python wrapper to link it to STT just like with KenLM.

I'm not saying it would be easy, but from my understanding, STT is made in such a way, it could be done with minimal modifications to the code base.

There is not much documentation on this (that I can find at least), so a good understanding of KenLM and it's implementation inside STT would greatly help.

**Reply by fquirin** [reply]
> Maybe the raspberry pi isn't the best platform for your needs.

It is the other way around. I'm trying to find the best ASR system for the Raspberry Pi not the best hardware for Coqui ;-)

> STT supports custom language models so technically nothing is stopping you to implement another LM with those desired traits.

As much as I'd like to spend countless hours on this I will leave this to the ASR experts with years of experience ^^.
Maybe there'll come a day where the [Hot-Word boost feature](https://stt.readthedocs.io/en/latest/HotWordBoosting-Examples.html) will actually become a "Hot-Sentence" boost 🙂

**Reply by HarikalarKutusu** [reply]
> Maybe there'll come a day where the [Hot-Word boost feature](https://stt.readthedocs.io/en/latest/HotWordBoosting-Examples.html) will actually become a "Hot-Sentence" boost

If you are working with small LMs with special command like structures, there is a trick: Glue the words... E.g.

"hey home, open the door" => "heyhomeopenthedoor", hot-word boost it and use it as an intent...

**Reply by fquirin** [reply]
That is actually working? Shouldn't that be out of vocabulary according to the docs: 

_"It is worth noting that boosting non-existent words in scorer (mostly proper nouns) or a word that share no phonetic prefix with other word in the input audio don’t change the final transcription."_

**Reply by HarikalarKutusu** [reply]
If you are creating your own scorer, you will have it in the language model. I applied the idea in my voice chess for the coordinates. E.g. instead of adding "alpha two" I added "alphatwo" and that made the recognition much better. I used a generator for that:

https://github.com/HarikalarKutusu/3d-voice-chess/tree/main/language-model-creation/chess-sentence-generators 

There are no hot-word boosting in that implementation though, there were too many to list.

I.e. you can also use: "heyhome" "openthedoor" like sub sentences, the first for wakeup, the second for intent.
